### ISSUE TITLE
Selected 320kbps Audio quality but the files are 305kbps

### ISSUE THREAD

**Issue by andrefromsbg** [jwallet/spy-spotify#516]
After adding this to Spytify.exe.config (menu dropdown did not work) I am able to record better sound quality:

```
<setting name="settings_media_bitrate_quality" serializeAs="String">
    <value>3</value>
</setting>
```

So with value 3 I selected MP3 320kbps on Audio quality but the files are all between 305 to 310kbps. Why is this?

**Comment by TraumaER** [comment]
Did you download the songs first? I'm thinking that Spotify itself is only streaming at that rate. I'm not sure how to go about confirming this though.

**Comment by andrefromsbg** [comment]
yes is also tried to download the songs first but this does not matter, files are not 320 but 305-310...

**Comment by jumph4x** [comment]
@andrefromsbg I have this issue as well, notably Pioneer gear hates non CBR (Constant BitRate). I looked briefly into how to configure the source code to use LAME CBR but did not find an easy way, so I actually have a re-encode step. 

Current presets use ABR/VBR, that's why. 

@TraumaER does not matter whatsoever what the bitrate of songs coming in is since Spy-Spotify uses local audio interface loopback. Meaning the digital stream is fully DAC'ed, then encoded again into chosen setting. 

The core problem is that CBR strategies are not available in the GUI. A trivial change the author could implement to save some of us a bunch of time.

**Comment by jumph4x** [comment]
Here is the source code area that shows the GUI settings indices and what LAME settings presets they tap into: https://github.com/jwallet/spy-spotify/blob/5a8f62c4c3144b1eebed9b90a3f544bfa2cea37f/EspionSpotify/frmEspionSpotify.cs#L458

**Comment by jumph4x** [comment]
Here is some documentation for LAME (unofficial): https://timothygu.me/lame/usage.html

**Comment by jumph4x** [comment]
Here is the wrapper this project uses to interface with the DLLs: https://github.com/Corey-M/NAudio.Lame

**Comment by jumph4x** [comment]
Looks like passing an integer will instruct NAudio wrappers to use CBR: 

![Image](https://github.com/user-attachments/assets/a823ed7f-314d-43cc-b0ae-8addb0754266)

**Comment by jumph4x** [comment]
@jwallet this suggests some way of adding custom bitrates, is there a way to simply add CBR320? 

https://github.com/jwallet/spy-spotify/blob/5a8f62c4c3144b1eebed9b90a3f544bfa2cea37f/EspionSpotify/frmEspionSpotify.cs#L476

**Comment by jumph4x** [comment]
<del>Here is where NAudio defines valid presets, none of which are CBR :'( </del>

https://github.com/Corey-M/NAudio.Lame/blob/master/LameDLLWrap/LAMEPreset.cs

EDIT: See discussion down below, potentially not true.

**Comment by jumph4x** [comment]
The wrapper does have a setting for VBRMode: https://github.com/Corey-M/NAudio.Lame/blob/07067e269439a166788d5fadad76adc60fb414b7/LameDLLWrap/VBRMode.cs#L44

(The hunt continues, haha).

**Comment by jumph4x** [comment]
This shows us that the LameMP3FileWriter accept either a preset/quality, a bitrate or an entire LameConfig: https://github.com/Corey-M/NAudio.Lame/blob/master/NAudio.Lame/LameMP3FileWriter.cs#L85-L129

**Comment by jumph4x** [comment]
LameConfig seems to confirm that if we initialize the writer using an integer to represent the BitRate, it will just use CBR: https://github.com/Corey-M/NAudio.Lame/blob/master/NAudio.Lame/LameConfig.cs#L147-L171

**Comment by jumph4x** [comment]
Actually, the legacy preset system may provide an entry point into CBR without ripping up this project:

![Image](https://github.com/user-attachments/assets/3639f05d-f19e-4f6f-9c2e-cd0a0fa2a31d) Source: https://wiki.hydrogenaud.io/index.php?title=LAME

This shows that the INSANE preset is a CBR strategy and this preset still exists for compatibility in the wrapper: https://github.com/Corey-M/NAudio.Lame/blob/master/LameDLLWrap/LAMEPreset.cs#L122

**Comment by jumph4x** [comment]
Settings reset script file gives us a clue where dotnet/VS store the app settings:

```
@ECHO OFF
ECHO Reset Spytify to its default settings.
ECHO If you continue, it will remove all user settings from any Spytify current or older versions on your system.
PAUSE
DEL /S /q "%userprofile%\AppData\Local\Spytify\*"
PAUSE
```

**Comment by jumph4x** [comment]
Looks like we did it, team. Quick solution for people like me:

- The settings are stored in `%userprofile%\AppData\Local\Spytify\Spytify.exe_Url_*/user.config` (look at app version and last modified date to understand which folder is the relevant one. If you are manually using File Explorer, the `%userprofile%` should resolve to something like `C:\Users\userprofile`, replace the last folder name with your account name.
- Once you open the file, find the setting line for `settings_media_bitrate_quality`, for me this was Line 8.
- Here, you have the freedom to use any LAMEPreset value from the Enum, here is a reference: https://github.com/Corey-M/NAudio.Lame/blob/master/LameDLLWrap/LAMEPreset.cs#L122

For my purposes, and purposes of the original author, you want to set this value to 1003 as that's the only (from my understanding, see above discussion) CBR preset (maintained for legacy/compatibility reasons). You can also any other value.

**Comment by jumph4x** [comment]
In v 1.12, it might not save bitrate settings correctly, so add this line manually to the file if that setting does not already exist there:

```
<setting name="settings_media_bitrate_quality"
                serializeAs="String">
                <value>1003</value>
            </setting>
```

**Comment by jumph4x** [comment]
If you are successful, the name of the custom LAMEPreset will be rendered to the GUI (cool consideration), it will look like this:

![Image](https://github.com/user-attachments/assets/6c17b9d0-1f3b-4fd9-806c-658713c73b0b)

**Comment by jumph4x** [comment]
Confirmed works: 

![Image](https://github.com/user-attachments/assets/04cc8188-d451-4a2c-8546-9d2d2d0699b7)

**Comment by jumph4x** [comment]
For discoverability: this should help with artifacts when mixing MP3 files on Pioneer XDJ-R1, XDJ-R2, XDJ-R3, XDJ-RR and potentially the larger standalone units when using the USB drive mode.

**Comment by jumph4x** [comment]
I've made a PR, you can download this and build it yourself using Visual Studio 2022: https://github.com/jwallet/spy-spotify/pull/520

It works, see screenshot.

**Comment by mehdijoe** [comment]
> Looks like we did it, team. Quick solution for people like me:
> 
> * The settings are stored in `%userprofile%\AppData\Local\Spytify\Spytify.exe_Url_*/user.config` (look at app version and last modified date to understand which folder is the relevant one. If you are manually using File Explorer, the `%userprofile%` should resolve to something like `C:\Users\userprofile`, replace the last folder name with your account name.
> * Once you open the file, find the setting line for `settings_media_bitrate_quality`, for me this was Line 8.
> * Here, you have the freedom to use any LAMEPreset value from the Enum, here is a reference: https://github.com/Corey-M/NAudio.Lame/blob/master/LameDLLWrap/LAMEPreset.cs#L122
> 
> For my purposes, and purposes of the original author, you want to set this value to 1003 as that's the only (from my understanding, see above discussion) CBR preset (maintained for legacy/compatibility reasons). You can also any other value.

What an amazing job! it works like a charm! and it is legit true CBR 320 kb
I am gonna save this edited user.config file somewhere safe! thank you!

![Image](https://github.com/user-attachments/assets/9145c427-8e28-41af-b955-42006cf90420)

**Comment by jumph4x** [comment]
Glad to be of help.
